Recipe crud functions (PT 1)

rae-stanton opened the pull request:
This allows a user to create, edit, and update a recipe. This also includes updated tests, since I updated the valid recipe requirements to have a short description in the model. Tests have been added for the edit button, that it renders the form correctly, and in an invalid scenario, it displays the appropriate error message. Deletion to come soon in a separate ticket as well as future styling for the new/edit/delete form.

rae-stanton committed:
Recipe CRUD 🚧

WIP on the recipe CRUD actions. This has placeholders for new/create in the recipe controller, updates the routes to contain *all* of the routes for recipes, and has a very simplified form helper (that currently doesn't do anything, but it WILL!!!). The form is in the recipe card partial, but has not been rendered as a partial in `new.html.erb` yet.

rae-stanton committed:
Create recipe WIP

This adds logic in the recipes controller for the new and create action and defines the minimum params for a recipe to be created properly. The form needs to reflect the params still - WIP

rae-stanton committed:
New recipe now works woo~

Recipe form has been updated - it actually works and redirects you to that recipe's show page!!!!!

rae-stanton committed:
Added edit recipe functionality!

This adds the edit/update CRUD actions to the recipe's show page. This update also adds the recipe form to the `recipecard` partial.

rae-stanton committed:
SPEC WIPS

Had to fix spec to meet new description criteria

rae-stanton committed:
Feature spec for recipe edit action

This finishes updating all specs for the updated recipe controller/model to reflect the description assertion (I know this is a bigger PR) I definitely need to do separate branches in these situations in the future. BUT this adds assertions for correct rendering of the edit page and for an invalid scenario with error message.

spaulsandhu commented on app/models/recipe.rb:
```diff
@@ -29,6 +29,7 @@ class Recipe < ApplicationRecord
   }
   validates :difficulty, presence: true
   validates :name, presence: true
+  validates :description, presence: true, length: { minimum: 10 }
```
Should we add a unit test for this? 🤔

rae-stanton replied:
Maybe a separate one sure! It's sort of in the feature spec now but could use a specific test

spaulsandhu replied:
Makes sense, lets do it in a separate PR!

spaulsandhu commented on spec/models/recipe_spec.rb:
```diff
@@ -20,7 +20,9 @@
 RSpec.describe Recipe, type: :model do
   let(:faust) { User.create(name: "Faust") }
-  let(:recipe) { Recipe.create(name: "Eggplant Hee Hee", author: faust, difficulty: "beginner", favorite: true) }
+  let(:recipe) { Recipe.create(name: "Eggplant Hee Hee", author: faust,
+                              difficulty: "beginner", favorite: true,
+                              description: "Something tasty") }
```
Rubocop could help with the formatting of these multi-line blocks.

spaulsandhu replied:
Additionally, you should look at building the habit of running rubocop on all your changed specs and `.rb` files before committing OR look into adding some kind of plugin for VS code that will automatically run rubocop for you!

spaulsandhu commented on app/views/recipes/_recipecard.html.erb:
```diff
@@ -0,0 +1,41 @@
+
```
I think it would make more sense and follow established rails paradigms/best practices if we re-name this file to `_form.html.erb` or `_recipe_form.html.erb`. When I see "recipecard" I think of the card that we use to display the recipe.

spaulsandhu commented on app/views/recipes/edit.html.erb:
```diff
@@ -0,0 +1,2 @@
+<h1>Edit Recipe</h1>
+<%= render "recipecard", recipe: @recipe %>
```
After we change the file name from `_recipecard.html.erb` to `_form.html.erb` we'll need to update this line.

spaulsandhu commented on app/views/recipes/new.html.erb:
```diff
@@ -0,0 +1,3 @@
+<h1>New Recipe</h1>
+
+<%= render "recipecard", recipe: @recipe %>
```
Same as above ^^^

spaulsandhu reviewed (changes requested):
This is phenomenal work @rae-stanton! The way you blasted through this slice of work was really impressive. I left a couple of comments about some Rubocop formatting/clean-up we could do and a suggestion for a file-name change that we should make. Once those changes are done, let me know and I'll re-review this PR.

Great work!

![well_done](https://media.giphy.com/media/sBLcw5Ic4QUTK/giphy.gif)

rae-stanton committed:
Suggested Changes

Ran RuboCop in Spec to make it look a little cleaner and made suggested name changes. I'll do some unit tests in a separate ticket, since they should be added in general.

rae-stanton committed:
Colored tags

Added for fun - quick change to add badges instead of plain text for recipe difficulties on show/index pages.

spaulsandhu reviewed (approved):
Looks great, thanks for the changes!
